src/lib/transaction_snark/transaction_snark.ml

```diff
@@ -999,38 +1061,94 @@ module Base = struct
+    in
     let%bind token_default =
       Token_id.(Checked.equal token (var_of_t default))
     in
     let%bind () =
       [%with_label "Validate tokens"]
-        (let%bind () =
-           (* TODO: Remove this check and update the transaction snark once we
-              have an exchange rate mechanism. See issue #4447.
-           *)
-           [%with_label "Validate fee token"]
-             (Token_id.Checked.Assert.equal fee_token
-                Token_id.(var_of_t default))
-         in
-         [%with_label "Validate delegated token is default"]
-           Boolean.(Assert.any [token_default; not is_stake_delegation]))
+        (Checked.all_unit
+           [ [%with_label "Fee token is valid"]
+               Boolean.(Assert.is_true (not fee_token_invalid))
+           ; [%with_label
+               "Fee token is default or command allows non-default fee"]
+               (Boolean.Assert.any
+                  [ fee_token_default
+                  ; is_payment
+                  ; is_stake_delegation
+                  ; is_fee_transfer ])
+           ; (* TODO: Remove this check and update the transaction snark once we
+               have an exchange rate mechanism. See issue #4447.
+            *)
+             [%with_label "Fees in tokens disabled"]
+               (Boolean.Assert.is_true fee_token_default)
+           ; [%with_label "Token is valid or command allows invalid token"]
+               Boolean.(Assert.any [not token_invalid; is_create_account])
+           ; [%with_label
+               "Token is default or command allows non-default token"]
+               (Boolean.Assert.any
+                  [ token_default
+                  ; is_payment
+                  ; is_create_account
+                    (* TODO: Enable this when fees in tokens are enabled. *)
+                    (*; is_fee_transfer*) ])
+           ; [%with_label
+               "Token is non-default or command allows default token"]
+               Boolean.(
+                 Assert.any
+                   [ not token_default
+                   ; is_payment
+                   ; is_stake_delegation
+                   ; is_fee_transfer
+                   ; is_coinbase ]) ])
+    in
+    let%bind creating_new_token =
+      Boolean.(is_create_account && token_invalid)
+    in
+    let%bind next_available_token, token =
+      (* TODO: Use next_available_token from the protocol state.  *)
+      let next_available_token = Token_id.(var_of_t (next default)) in
+      let%bind () =
+        [%with_label "New token creation is disabled"]
+          Boolean.(Assert.is_true (not creating_new_token))
+      in
+      let%bind token =
+        Token_id.Checked.if_ creating_new_token ~then_:next_available_token
+          ~else_:token
+      in
+      let%map next_available_token =
+        Token_id.Checked.next_if next_available_token creating_new_token
+      in
+      (next_available_token, token)
+    in
+    ignore next_available_token ;
```

Comment:
Do we need to bump the value in the protocol state somehow?

Reply:
Now I think about it, yes we do. I'll revive #5169, thanks.